### ISSUE TITLE
Implementation Help!!!

### ISSUE THREAD

**Issue by liuzhi136** [apache/incubator-mxnet#8720]
I'm trying to implement SAN model proposed in "Stacked Attention Network for Image Question Answering" in mxnet, which can be download below. 

[Yang_Stacked_Attention_Networks_CVPR_2016_paper.pdf](https://github.com/apache/incubator-mxnet/files/1486170/Yang_Stacked_Attention_Networks_CVPR_2016_paper.pdf)

My implementations looks like below :
[san_vqamodel_with_gluon.txt](https://github.com/apache/incubator-mxnet/files/1486193/san_vqamodel_with_gluon.txt)

I had downloaded the DAQUAR dataset and use the reduced version to reproduce the result of this paper. I think the model described clearly enough in this paper. However, this model I implement doesn't converge at all and always fluctuate after several epochs, and its training and validation error less than 3. I've tried various hyperparameter sets to adjust the training procedure and does not work out. It's urgent problem for me, and I really need help to solve it as soon as possible. I really want to know if my implementation is correct. Does any could help me? Cause it's a really urgent. 
Any help will be appreciated!

**Comment by liuzhi136** [comment]
@mli @piiswrong @szha

**Comment by szha** [comment]
What have you tried to isolate the problem?

**Comment by liuzhi136** [comment]
@szha I do not understand the meaning of "isolate the problem". Can you explain it? And Is it OK for you that I have your Wechat or email?

**Comment by szha** [comment]
I'm not familiar with this model, so I'm trying to help by offering from the perspective of how I would approach it.

Have you tried anything to find which parts of the whole pipeline might have problems? For example, are you sure your data pipeline is working correctly? Have you examined samples from your iterator to see if it looks correct? If just using a small subset as training set, does your model overfit to the set?

**Comment by liuzhi136** [comment]
I just examine the data batch I provide to the model. one of batches looks like below and the next one is its corresponding text version. I can assure you that my data pipeline is working correctly. The data set I use to train the model is the same as used in this paper, and I just want to reproduce this experiment result see whether it really work. I think it is not overfitting cause the training loss and validations loss are both high. I think it's still underfitting. Can you just read this paper to check if my model structure is correct? Cause it is not too complicated and won't take you too much time. maybe about ten minutes at most.

![image](https://user-images.githubusercontent.com/13534043/33006782-a80a5e0e-ce06-11e7-8af3-e6cf6934d987.png)

![image](https://user-images.githubusercontent.com/13534043/33006123-19511100-ce04-11e7-9867-7e856d59255c.png)

**Comment by liuzhi136** [comment]
I still wander whether the way I pad end character is correct ?

**Comment by szha** [comment]
I'm asking you to try and overfit a small dataset first as sanity check. If your model can't overfit a small subset of the data, then your model is wrong. This is the first step.

Unfortunately I can't read the paper and reproduce this for you right now.

**Comment by szha** [comment]
From the screenshot it looks like the indexing of the text is correct. Do the questions match the input images?

**Comment by szha** [comment]
For the text part, you're using end-of-sentence token as the padding, and are unrolling all the way even for paddings. Usually you would want to 1) use a separate token for padding, and 2) minimize the impact of padding by including as little of it as possible. Having these long paddings can impact your text model performance.

**Comment by liuzhi136** [comment]
I see. Thank you for helping me a lot. My model can overfit this data set. I had record my training loss, it decreases all the time during training process goes, the validation loss however increases after dozens of epochs. I think this is the clue of overfitting. Also, my image feature matches the questions. 
For second way for padding, do you mean I should use bucketing to try it again?
I do not clearly understand the first way you said " use a separate token for padding". What's that mean?

**Comment by szha** [comment]
The '<eos>' and '<pad>' can be two separate tokens, because they bear different meanings. By using a separate token, your model can learn a new embedding for padding that's different from those of end-of-sentence markers.

Yes, bucketing can indeed help reduce the padding masks.

**Comment by liuzhi136** [comment]
How to do the "separate token"? This really is my first time hearing about this trick. Could you explain more detail over how to do separate token ""? I really need to solve this problem which I have struggled for several weeks!!  

Thank you so much!

**Comment by szha** [comment]
simple, you can just pad the sequences with '<pad>' tokens and put only one '<eos>' before the padding tokens.

**Comment by liuzhi136** [comment]
Let me pose an example, 
say I have a question like "What is found on the floor in image1327 ?" and bptt=20, 
then I should turn it into "What is found on the floor in image1327 ? tokens and put only one ". 
After that, padding with "<eos>" it looks like  below
"What is found on the floor in image1327 ? tokens and put only one <eos><eos><eos><eos><eos><eos>". 
Am I right?

**Comment by szha** [comment]
```
What is found on the floor in image1327 ? tokens and put only one <eos> <pad> <pad> <pad> <pad> <pad>
```

**Comment by liuzhi136** [comment]
These words as separated tokens, I think, will bring some noisy word infos into hidden vector produced by LSTM structure instead of padding tokens that bring some meaningless infos. Does this form will improve the performance? I don't have any  idea about it. Anyway, I will try it. But I'm still confused its theory or mechanism behind this trick. Do you have any idea about it?
I still noticed that the last five padding tokens. Is this padding token still in the vocabulary set?
say "word2idx={'<pad>':0, '<eos>':1}"

**Comment by liuzhi136** [comment]
@szha May I ask you another question? I just noticed the symbol API SoftmaxOutput function has two parameters "ignore_label=_Null" and "use_ignore=_Null" which I think may very helpful when I training models, I didn't however find these two parameters in the Gluon Loss API "SoftmaxCrossEntropyLoss" function. So How can I do the same thing as the symbol API does in gluon?

**Comment by indhub** [comment]
@liuzhi136 This is not an 'issue'. For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) is a good place to get help.
